SW801F20/P8
# Feature bluetooth sync

### Feature Description
This PR implements/solves the following two trello tasks:
[Post location dead reckoning + API calculate coordinates](https://trello.com/c/qy0bvcUH)
[Bluetooth Synchronization](https://trello.com/c/4anFEl9x)

This features implements functionality that moves two devices (friends) closer to each other when in Bluetooth range and certain thresholds (criteria) are fulfilled. 

It also implements an endpoint in the API that allows a client to update their GPS location by sending a distance and orientation to the API. In this way, the functionality of calculating a new GPS location based on dead reckoning is done serverside. 

### Most Relevant Files
#### APP
Location_DAO
RSSI_DAO
User_DAO
LocationRepository
UserRepository
RSSIRepository
MainActivity
MapsViewModel
RSSIViewModel

#### API
DeadCrumbsApi.cs
LocationService
UserService

### How To Run
- Internet is required
- Add a valid API key to google_maps_api.xml
- Open Project with Android Studio
- Requires a Android Phone connected (and with developer mode enabled)
- Run as debug on Android

### How To Test - DeadReckoning
- Create a new user using the API (with unique name and mac address)
- Change the hardcoded my_username in the MainActivity companion object
- Run the APP
- Open the map
- Take a step (or more than one)
- Check to see that both the arrow moves in the Map and that the location changes in the DB using the [swagger API](http://130.225.57.95:8393/swagger/index.html) GET /Location/{username} endpoint.

### How To Test - Bluetooth Synchronization
- Requires two devices (second one can be any bluetooth device that you can get the mac address from)
- Create two users using the API (with unique name and correct BLUETOOTH mac address)
- Post a new location to each of your two users (two distinct locations)
- Change the hardcoded my_username in the MainActivity companion object to the user corresponding to your phone
- Run the APP
- Place both devices within 2 meters and with no obstacles.
- Check to see that both the arrow and the second users location sync up to be placed at the same location (the location should be exactly in the middle of the two previous locations)

Note: We have done a huge merge with 7+ conflicting files, so please pay attention to any defects in previously implemented functionality as this may have been the result of a bad merge
